File: himark/parser/_count.py
"""Count-modifier parsing: a `[count]` string → a typed `CountSpec`.

Standalone from the σ-grammar resolution in phase3 — it operates purely on the
count string between the brackets (`"1..6"`, `"a,b,c"`, `"#i"`), with no
knowledge of the universe the count applies to.
"""

import re

from himark.models import nodes_typed as t
from himark.models.exceptions import CompileError

_COUNTREF = re.compile(r"#(\d+)")


def parse_count(src: str) -> t.CountSpec:
    """Parse a count modifier string into a count descriptor.

    Forms: `[n]`, `[x..]`, `[..y]`, `[x..y]`, `[x..y..s]` (stride),
    `[a,b,c]` (union), `[#i]` (count-reference)."""
    src = src.strip()
    # `[#i]` — repeat exactly group i's repetition count (resolved at match time).
    m = _COUNTREF.fullmatch(src)
    if m:
        return t.CountRefSpec(group=int(m.group(1)))
    # `[a,b,c]` — an explicit union of exact counts.
    if "," in src:
        try:
            values = sorted({int(p.strip()) for p in src.split(",")})
        except ValueError:
            raise CompileError(f"Invalid count expression: [{src}]") from None
        return t.CountSet(values=values)
    # `[n]` / `[x..y]` with optional stride `..s`.
    m = re.fullmatch(r"(\d*)(?:\.\.(\d*)(?:\.\.(\d+))?)?", src)
    if not m or not (m.group(1) or ".." in src):
        raise CompileError(f"Invalid count expression: [{src}]")
    lo, hi, step = m.groups()
    if ".." not in src:  # exact [n]
        return t.CountRange(min=int(lo), max=int(lo))
    step_n = int(step) if step else 1
    max_n = int(hi) if hi else None
    if step_n != 1 and max_n is None:
        raise CompileError(f"A strided count needs an upper bound: [{src}]")
    return t.CountRange(min=int(lo) if lo else 0, max=max_n, step=step_n)

File: rust/src/parser/count.rs
use crate::parser::nodes::CountSpec;

/// Parse a count modifier string (the content between `[` and `]`) into a
/// `CountSpec`. Mirrors himark/parser/_count.py.
pub fn parse_count(src: &str) -> Result<CountSpec, String> {
    let src = src.trim();

    // [#i] — count-reference: repeat exactly as many times as group i.
    if let Some(rest) = src.strip_prefix('#') {
        if !rest.is_empty() && rest.chars().all(|c| c.is_ascii_digit()) {
            let group: usize = rest
                .parse()
                .map_err(|_| format!("Invalid count expression: [{src}]"))?;
            return Ok(CountSpec::Ref { group });
        }
    }

    // [a,b,c] — explicit union of exact counts.
    if src.contains(',') {
        let mut values: Vec<usize> = src
            .split(',')
            .map(|p| {
                p.trim()
                    .parse::<usize>()
                    .map_err(|_| format!("Invalid count expression: [{src}]"))
            })
            .collect::<Result<Vec<_>, _>>()?;
        values.sort_unstable();
        values.dedup();
        return Ok(CountSpec::Set { values, group: None });
    }

    // [n] or [x..y] (x and y each optional).
    if src.contains("..") {
        let idx = src.find("..").unwrap();
        let lo_str = &src[..idx];
        let hi_str = &src[idx + 2..];
        let min: usize = if lo_str.is_empty() {
            0
        } else {
            lo_str
                .parse()
                .map_err(|_| format!("Invalid count expression: [{src}]"))?
        };
        let max: Option<usize> = if hi_str.is_empty() {
            None
        } else {
            Some(
                hi_str
                    .parse()
                    .map_err(|_| format!("Invalid count expression: [{src}]"))?,
            )
        };
        return Ok(CountSpec::Range { min, max, group: None });
    }

    // [n] — exact count.
    if src.chars().all(|c| c.is_ascii_digit()) && !src.is_empty() {
        let n: usize = src
            .parse()
            .map_err(|_| format!("Invalid count expression: [{src}]"))?;
        return Ok(CountSpec::Range { min: n, max: Some(n), group: None });
    }

    Err(format!("Invalid count expression: [{src}]"))
}
